### PR TITLE
Task/109 secure pin functionality

### PR DESCRIPTION
**Pull request by storkandstars**
# Overview

Task implements application protection via a security secret that is generated on account creation. This secret is encrypted with a user defined pin. We use a AES encryption algo that generates and stores a unique iv number.

There was a large migration from a Navigator to Router. Implementation of Router is via [go_router](https://gorouter.dev/).

#### 1. Creating a security pin

- When there is an initial login on a device a user is redirected to a pin setup screen.
- User inserts 4 digit pin
- A square submit button appears
- After submit a user is redirected to a success screen (Both these screens cannot be left by clicking on back button or lading back)
- After clicking on `start using app` user enters the dashboard.
- During setting up a security pin, a user is always returned to the setup pin screen after he leaves the entire app.

#### 2. Verifying a security pin

- When a security pin is set on a device the user is asked to enter the pin on app enter.
- if the pin is correct, the user is brought into dashboard when the form gets 4 digit pin
- if the user enters incorrect pin, a remaining attempts message appears under pin input
- if the user leaves the app and comes back, the remaining attempts appears with the last seen value
- After correct pin enter, remaining attempts are reset to the initial value (that is by default 6)
- If the user enters 6 times pin incorrecly, he is log out.
 
#### 3. Protected screens

- going to the settings/configuration screen there is a protected screen `password change`. After going to that screen, the user should be asked to verify his pin.
- If the pin is 6 time incorrectly entered, the user is pushed to the previous screen

### CODE REVIEW

File: lib/view/screens/application_screen.dart [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                      TapBarItem(
+                          key: const ValueKey(AuthenticatedRouter.dashboard),
+                          tapBarItemType: TapTabItemType.button,
+                          icon: Icons.home_outlined,
+                          activeIcon: Icons.home,
+                          label: context.locale.homeTapBarLabel,
+                          onTap: (context, key) {
+                            GoRouter.of(context)
+                                .goNamed(AuthenticatedRouter.dashboard);
+                            // ApplicationRoute.key.currentState
```
Comment (jvrastil) [inline lib/view/screens/application_screen.dart:58]: remove unused code please

File: lib/view/screens/dashboard_screen.dart [hunk 2]
```diff
@@ -29,8 +30,9 @@ class DashboardScreen extends StatelessWidget {
               ),
               child: InkWell(
                 onTap: () {
-                  AuthenticatedRoute.key.currentState
-                      ?.pushReplacementNamed(AuthenticatedRoute.inventory);
+                  GoRouter.of(context).goNamed(AuthenticatedRouter.inventory);
+                  // ApplicationRoute.key.currentState
```
Comment (jvrastil) [inline lib/view/screens/dashboard_screen.dart:34]: remove unused code here as well please

File: lib/view/screens/portfolio_screen.dart [hunk 3]
```diff
@@ -45,11 +44,12 @@ class PortfolioScreen extends StatelessWidget {
                       return WalletListItem(
                         key: Key(wallet.assetId),
                         assetWallet: wallet,
-                        onTap: () =>
-                            AuthenticatedRoute.key.currentState?.pushNamed(
-                          AuthenticatedRoute.wallet,
-                          arguments: wallet,
-                        ),
+                        onTap: () {
+                          //   ApplicationRoute.key.currentState?.pushNamed(
```
Comment (jvrastil) [inline lib/view/screens/portfolio_screen.dart:48]: should't we redirect user with new router or is it prepared for some future task?

File: lib/view/widgets/keypad.dart [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+              Row(
+                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+                children: [
+                  for (var item in row.children)
+                    Expanded(
+                      flex: stretchItems ? 1 : 0,
+                      child: item,
+                    )
+                ],
+                // children: row.children,
```
Comment (jvrastil) [inline lib/view/widgets/keypad.dart:57]: remove

File: lib/view/widgets/keypad.dart [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    this.onProceed,
+    this.onErase,
+    this.displayProceedButton = false,
+    this.displayEraseButton = false,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return _KeypadGrid.builder(
+      // stretchItems: true,
```
Comment (jvrastil) [inline lib/view/widgets/keypad.dart:85]: remove this as well please

File: lib/view/widgets/pin_digits.dart [hunk 6]
```diff
@@ -14,9 +13,9 @@ class PinDigits extends StatefulWidget {
   const PinDigits({
     super.key,
     this.pin,
-    this.pinLength = SecurityFormState.pinLength,
+    this.pinLength = ValidatedPin.length,
     this.pinDigitStyle = PinDigitStyle.hideAllAfterTime,
-    this.duration = const Duration(milliseconds: 1000),
+    this.duration = const Duration(milliseconds: 0),
```
Comment (jvrastil) [inline lib/view/widgets/pin_digits.dart:18]: is this supposed to be 0 when default style is `hideAllAfterTime` ?

File: lib/view/widgets/pin_digits.dart [hunk 7]
```diff
@@ -86,12 +85,10 @@ class _PinDigitsState extends State<PinDigits> {
...
     }
 
-    return Column(children: [
-      Row(
-        mainAxisAlignment: MainAxisAlignment.center,
-        children: pinNumbers,
-      ),
-    ]);
+    return Wrap(
+      // mainAxisAlignment: MainAxisAlignment.center,
```
Comment (jvrastil) [inline lib/view/widgets/pin_digits.dart:89]: you can remove this as well

**Review by jvrastil — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 8]
```diff
@@ -48,11 +48,6 @@ app.*.map.json
 # used in gitlab workflows
 version.yml
 
-## generated files
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply .gitignore:51]: There was a duplicity in comments for generated files. At the bottom of the gitignore file there is definition for ignoring generated files provided from PR.

File: lib/data/model/account_wallet.dart [hunk 9]
```diff
@@ -92,13 +92,15 @@ class LockableSeed with AesEncryption {
 
   String unlock(String key) {
     if (!_isLocked) return _mnemonic;
-    return decrypt(key, _mnemonic);
+    // TODO - Change IV value!
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/data/model/account_wallet.dart:95]: This section will we changed in issue #116 where we implement seed encryption.

File: lib/data/repositories/auth_repository.dart [hunk 10]
```diff
@@ -92,45 +86,21 @@ class AuthRepository with JwtMixin {
           accessToken: body.accessToken,
         ));
 
-        final userId = decodedIdToken['name'];
+        final userId = decodedIdToken['sub'];
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/data/repositories/auth_repository.dart]: Yes. It was fixed. [0e90f10](https://github.com/SimplioOfficial/simplio-app/pull/121/commits/0e90f10443baaf60297e7746d8f4a026f8146fa4)

File: lib/view/routes/guards/protected_guard.dart [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                        maintainSize: true,
+                        maintainAnimation: true,
+                        maintainInteractivity: true,
+                        maintainSemantics: true,
+                        maintainState: true,
+                        child: Text(
+                          attempts > 1
+                              ? "${context.locale.protectedErrorRemaining} $attempts"
+                              : context.locale.protectedErrorLast,
+                          style: const TextStyle(color: Colors.red),
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/view/routes/guards/protected_guard.dart]: It was fixed. [e1a34bd](https://github.com/SimplioOfficial/simplio-app/pull/121/commits/e1a34bd3319fe73819cdeb9a1ff9469d62e21d73)

File: lib/logic/cubit/sign_up_form/sign_up_form_state.dart [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  }) {
+    return SignUpFormState._(
+      login: ValidatedEmail(value: login ?? this.login.toString()),
+      password: ValidatedPassword(value: password ?? this.password.toString()),
+      response: response,
+    );
+  }
+}
+
+// class SignInForm extends Equatable {
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/logic/cubit/sign_up_form/sign_up_form_state.dart:43]: It was fixed. [1a58353](https://github.com/SimplioOfficial/simplio-app/pull/121/commits/1a58353109d756b42fd57a71835178b080249dfb)

File: lib/view/authenticated_app.dart [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+          )..loadAccount(accountId),
+        ),
+        BlocProvider(create: (_) => TapBarCubit()),
+      ],
+      child: Builder(
+        builder: (context) {
+          final r = AuthenticatedRouter.of(context).router;
+          return Builder(
+            builder: (context) {
+              // _setSystemUIOverlayStyle(context);
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/view/authenticated_app.dart]: Yes. By separating build context from router state. We are able to keep state of the router and remember screens on rebuild.
Comments are removed here [dda3c4b](https://github.com/SimplioOfficial/simplio-app/pull/121/commits/dda3c4b3e16a38b9dbf694a7d4ff437629066d9d)

File: lib/view/widgets/pin_digits.dart (hunk 6 above)
Comment (storkandstars) [reply lib/view/widgets/pin_digits.dart]: Returned to the previous value [c4472b6](https://github.com/SimplioOfficial/simplio-app/pull/121/commits/c4472b6e36c52ecdb4a67fb422794ad1739cdae5)

File: lib/view/widgets/pin_digits.dart (hunk 7 above)
Comment (storkandstars) [reply lib/view/widgets/pin_digits.dart]: Removed

File: lib/view/widgets/keypad.dart (hunk 5 above)
Comment (storkandstars) [reply lib/view/widgets/keypad.dart]: Removed.

File: lib/view/widgets/keypad.dart (hunk 4 above)
Comment (storkandstars) [reply lib/view/widgets/keypad.dart]: Removed.

File: lib/view/screens/portfolio_screen.dart (hunk 3 above)
Comment (storkandstars) [reply lib/view/screens/portfolio_screen.dart]: We can make an task with screen changes due to ux in future task.

File: lib/view/screens/dashboard_screen.dart (hunk 2 above)
Comment (storkandstars) [reply lib/view/screens/dashboard_screen.dart]: Removed.

File: lib/view/screens/application_screen.dart (hunk 1 above)
Comment (storkandstars) [reply lib/view/screens/application_screen.dart]: Removed.

File: lib/view/routes/authenticated_router.dart [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      ),
+      observers: [
+        TapBarObserver.of(context),
+      ],
+      routes: [
+        GoRoute(
+          path: '/in',
+          redirect: (state) {
+            final account = context.read<AccountCubit>().state.account;
+            // final account = context.watch<AccountCubit>().state.account;
```
[the remark this answers is not in the document]
Comment (storkandstars) [reply lib/view/routes/authenticated_router.dart]: Removed.

**Review by jvrastil — APPROVED** [review]
looks good now 👍

**Pull request merged**